File: gtci/07-stacks/cpp/1-balanced-parentheses.cpp
#include <cassert>
#include <iostream>
#include <stack>

using namespace std;

class Solution {
   public:
    bool isValid(string s) {
        std::stack<char> stack;

        for (char ch : s) {
            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else {
                if (stack.size() == 0) return false;

                const char top = stack.top();
                stack.pop();

                if (ch == ')' && top != '(')
                    return false;
                else if (ch == ']' && top != '[')
                    return false;
                else if (ch == '}' && top != '{')
                    return false;
            }
        }

        return stack.size() == 0;
    }
};

int main() {
    Solution sol;

    assert(sol.isValid("()") == true);
    assert(sol.isValid("()[]{}") == true);
    assert(sol.isValid("(]") == false);
    assert(sol.isValid("([)]") == false);
    assert(sol.isValid("{[]}") == true);
    assert(sol.isValid("}") == false);

    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/07-stacks/python3/1-balanced-parentheses.py
class Solution:
    def isValid(self, s):
        stack = []

        for char in s:
            if char == "(" or char == "[" or char == "{":
                stack.append(char)
            else:
                if len(stack) == 0:
                    return False

                top = stack.pop()
                if char == ")" and top != "(":
                    return False
                elif char == "]" and top != "[":
                    return False
                elif char == "}" and top != "{":
                    return False

        return len(stack) == 0


if __name__ == "__main__":
    sol = Solution()

    assert sol.isValid("()") == True
    assert sol.isValid("()[]{}") == True
    assert sol.isValid("(]") == False
    assert sol.isValid("([)]") == False
    assert sol.isValid("{[]}") == True
    assert sol.isValid("}") == False

    print("All test cases passed.")
